File: nanopay/src/net/nanopay/tx/AuthenticatedAccountDAO.java
package net.nanopay.tx;

import foam.core.FObject;
import foam.core.X;
import foam.dao.DAO;
import foam.dao.ProxyDAO;
import foam.dao.Sink;
import foam.mlang.order.Comparator;
import foam.mlang.predicate.Predicate;
import foam.nanos.auth.AuthService;
import foam.nanos.auth.User;
import java.security.AccessControlException;

import foam.util.SafetyUtil;
import net.nanopay.account.CurrentBalance;
import static foam.mlang.MLang.EQ;

public class AuthenticatedAccountDAO
  extends ProxyDAO
{
  public final static String GLOBAL_BALANCE_CREATE = "balance.create.x";
  public final static String GLOBAL_BALANCE_READ = "balance.read.x";
  public final static String GLOBAL_BALANCE_UPDATE = "balance.update.x";
  public final static String GLOBAL_BALANCE_DELETE = "balance.delete.x";

  public AuthenticatedAccountDAO(X x, DAO delegate) {
    setX(x);
    setDelegate(delegate);
  }

  @Override
  public FObject put_(X x, FObject obj) {
    User user = (User) x.get("user");
    CurrentBalance currentBalance = (CurrentBalance) obj;
    AuthService auth = (AuthService) x.get("auth");

    if ( user == null ) {
      throw new AccessControlException("User is not logged in");
    }

    // if current user doesn't have permissions to create or update, force currentBalance's owner to be current user id
    if ( currentBalance.getId() == 0 || ! auth.check(x, GLOBAL_BALANCE_CREATE) || ! auth.check(x, GLOBAL_BALANCE_UPDATE) ) {
      currentBalance.setId(user.getId());
    }

    return super.put_(x, obj);
  }

  @Override
  public FObject find_(X x, Object id) {
    User        user = (User) x.get("user");
    AuthService auth = (AuthService) x.get("auth");

    if ( user == null ) {
      throw new AccessControlException("User is not logged in");
    }

    // fetch currentBalance from delegate and verify user either owns the currentBalance or has global read access
    CurrentBalance currentBalance = (CurrentBalance) getDelegate().find_(x, id);
    if ( currentBalance != null && ! SafetyUtil.equals(currentBalance.getId(), user.getId()) && ! auth.check(x, GLOBAL_BALANCE_READ) ) {
      return null;
    }

    return currentBalance;
  }

  @Override
  public Sink select_(X x, Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
    User user = (User) x.get("user");
    AuthService auth = (AuthService) x.get("auth");

    if ( user == null ) {
      throw new AccessControlException("User is not logged in");
    }

    boolean global = auth.check(x, GLOBAL_BALANCE_READ);
    DAO dao = global ? getDelegate() : getDelegate().where(EQ(CurrentBalance.ID, user.getId()));
    return dao.select_(x, sink, skip, limit, order, predicate);
  }

  @Override
  public FObject remove_(X x, FObject obj) {
    User user = (User) x.get("user");
    CurrentBalance currentBalance = (CurrentBalance) obj;
    AuthService auth = (AuthService) x.get("auth");

    if ( user == null ) {
      throw new AccessControlException("User is not logged in");
    }

    if ( currentBalance != null && ! SafetyUtil.equals(currentBalance.getId(), user.getId()) && ! auth.check(x, GLOBAL_BALANCE_DELETE) ) {
      throw new RuntimeException("Unable to delete bank account");
    }

    return super.remove_(x, obj);
  }

  @Override
  public void removeAll_(X x, long skip, long limit, Comparator order, Predicate predicate) {
    User user = (User) x.get("user");
    AuthService auth = (AuthService) x.get("auth");

    if ( user == null ) {
      throw new AccessControlException("User is not logged in");
    }

    boolean global = auth.check(x, GLOBAL_BALANCE_DELETE);
    DAO dao = global ? getDelegate() : getDelegate().where(EQ(CurrentBalance.ID, user.getId()));
    dao.removeAll_(x, skip, limit, order, predicate);
  }
}

File: nanopay/src/net/nanopay/account/AuthenticatedAccountDAO.js
/**
 * NANOPAY CONFIDENTIAL
 *
 * [2020] nanopay Corporation
 * All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of nanopay Corporation.
 * The intellectual and technical concepts contained
 * herein are proprietary to nanopay Corporation
 * and may be covered by Canadian and Foreign Patents, patents
 * in process, and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from nanopay Corporation.
 */

foam.CLASS({
  package: 'net.nanopay.account',
  name: 'AuthenticatedAccountDAO',
  extends: 'foam.dao.ProxyDAO',

  javaImports: [
    'foam.core.FObject',
    'foam.core.PropertyInfo',
    'foam.core.X',
    'foam.dao.DAO',
    'foam.dao.Sink',
    'foam.mlang.order.Comparator',
    'foam.mlang.predicate.Predicate',
    'foam.nanos.auth.*',
    'net.nanopay.contacts.PersonalContact',

    'static foam.mlang.MLang.EQ',
    'static foam.mlang.MLang.OR'
  ],

  constants: [
    {
      name: 'GLOBAL_ACCOUNT_READ',
      type: 'String',
      value: 'account.read.*'
    },
    {
      name: 'GLOBAL_ACCOUNT_UPDATE',
      type: 'String',
      value: 'account.update.*'
    },
    {
      name: 'GLOBAL_ACCOUNT_DELETE',
      type: 'String',
      value: 'account.remove.*'
    },
    {
      name: 'GLOBAL_ACCOUNT_CREATE',
      type: 'String',
      value: 'account.create'
    }
  ],

  messages: [
    { name: 'UPDATE_PERMISSION_ERROR_MSG', message: 'You do not have permission to update that account' },
    { name: 'CREATE_PERMISSION_ERROR_MSG', message: 'You do not have permission to create an account for another user'},
    { name: 'DELETE_PERMISSION_ERROR_MSG', message: 'Unable to delete bank account due to insufficient permissions' }
  ],

  axioms: [
    {
      name: 'javaExtras',
      buildJavaClass: function(cls) {
        cls.extras.push(`
          public AuthenticatedAccountDAO(X x, DAO delegate) {
            setX(x);
            setDelegate(delegate);
          }
        `
        );
      }
    }
  ],

  methods: [
    {
      name: 'put_',
      javaCode: `
        User user = ((Subject) x.get("subject")).getUser();
        Account newAccount = (Account) obj;
        AuthService auth = (AuthService) x.get("auth");
        DAO userDAO_ = (DAO) x.get("bareUserDAO");

        if ( user == null ) {
          throw new AuthenticationException();
        }

        Account oldAccount = (Account) getDelegate().find_(x, obj);
        boolean isUpdate = oldAccount != null;

        if ( isUpdate ) {
          boolean ownsAccount = newAccount.getOwner() == user.getId() && oldAccount.getOwner() == user.getId();
          // TODO: explicitly check for update on status, verifiedBy for admin
          if ( ! isUpdateDefault(oldAccount, newAccount) &&
               ! auth.check(x, GLOBAL_ACCOUNT_UPDATE) ) {
            throw new AuthorizationException("User can update only isDefault property");
          }

          if (
            ! ownsAccount &&
            ! auth.check(x, GLOBAL_ACCOUNT_UPDATE) &&
            ! ownsContactThatOwnsAccount(x, newAccount) &&
            ! ownsContactThatOwnsAccount(x, oldAccount)
          ) {
            throw new AuthorizationException(UPDATE_PERMISSION_ERROR_MSG);
          }
        } else if (
          newAccount.getOwner() != user.getId() &&
          ! auth.check(x, "account.create") &&
          ! ownsContactThatOwnsAccount(x, newAccount)
        ) {
          throw new AuthorizationException(CREATE_PERMISSION_ERROR_MSG);
        }

        return super.put_(x, obj);
      `
    },
    {
      name: 'find_',
      javaCode: `
        User user = ((Subject) x.get("subject")).getUser();
        AuthService auth = (AuthService) x.get("auth");

        if ( user == null ) {
          throw new AuthenticationException();
        }

        Account account = (Account) super.find_(x, id);

        if ( account == null ) return null;

        if (
          account.getOwner() == user.getId() ||
          auth.check(x, GLOBAL_ACCOUNT_READ) ||
          ownsContactThatOwnsAccount(x, account)
        ) {
          return account;
        }

        return null;
      `
    },
    {
      name: 'select_',
      javaCode: `
        User user = ((Subject) x.get("subject")).getUser();
        AuthService auth = (AuthService) x.get("auth");

        if ( user == null ) {
          throw new AuthenticationException();
        }

        boolean global = auth.check(x, GLOBAL_ACCOUNT_READ);
        DAO dao = global ? getDelegate() : getDelegate().where(
          OR(
            EQ(Account.OWNER, user.getId()),
            EQ(Account.CREATED_BY, user.getId())
          )
        );
        return dao.select_(x, sink, skip, limit, order, predicate);
      `
    },
    {
      name: 'remove_',
      javaCode: `
        User user = ((Subject) x.get("subject")).getUser();
        Account account = (Account) obj;
        AuthService auth = (AuthService) x.get("auth");

        if ( user == null ) {
          throw new AuthenticationException();
        }

        if (
          account != null &&
          account.getOwner() != user.getId() &&
          ! auth.check(x, GLOBAL_ACCOUNT_DELETE) &&
          ! ownsContactThatOwnsAccount(x, account)
        ) {
          throw new AuthorizationException(DELETE_PERMISSION_ERROR_MSG);
        }

        return super.remove_(x, obj);
      `
    },
    {
      name: 'removeAll_',
      javaCode: `
        User user = ((Subject) x.get("subject")).getUser();
        AuthService auth = (AuthService) x.get("auth");

        if ( user == null ) {
          throw new AuthenticationException();
        }

        boolean global = auth.check(x, GLOBAL_ACCOUNT_DELETE);
        DAO dao = global ? getDelegate() : getDelegate().where(EQ(Account.OWNER, user.getId()));
        dao.removeAll_(x, skip, limit, order, predicate);
      `
    },
    {
      name: 'ownsContactThatOwnsAccount',
      type: 'Boolean',
      args: [
        { type: 'Context', name: 'x' },
        { type: 'Account', name: 'account' }
      ],
      documentation: `
        Check if the user in the context owns a contact that owns the given account.
        @param x The user context.
        @param account The account to check.
        @return true if the given account is owned by a contact that the user owns.
      `,
      javaCode: `
        User user = ((Subject) x.get("subject")).getUser();
        User owner = account.findOwner(x);
        return owner instanceof PersonalContact && ((PersonalContact) owner).getOwner() == user.getId();
      `
    },
    {
      name: 'isUpdateDefault',
      type: 'Boolean',
      args: [
        { type: 'Account', name: 'oldAccount' },
        { type: 'Account', name: 'newAccount' }
      ],
      javaCode: `
        String ignore = "isDefault, name, description";
        FObject nu = (FObject) newAccount.fclone();
        FObject old = (FObject) oldAccount.fclone();
        for ( String propName : ignore.split("\\\\s*,\\\\s*") ) {
          PropertyInfo prop = (PropertyInfo) nu.getClassInfo().getAxiomByName(propName);
          if (prop != null) {
            prop.clear(nu);
            prop.clear(old);
          }
        }
        return nu.equals(old);
      `
    }
  ]
});

